sw/device/silicon_creator/lib/drivers/otp.c

```diff
@@ -0,0 +1,77 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+
+#include <stddef.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "otp_ctrl_regs.h"
+
+static rom_error_t otp_read_start(const otp_t *otp, uint32_t address) {
+  uint32_t status =
+      mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+  // Check if otp_ctrl is busy.
+  if (!bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT)) {
+    return kErrorOtpBusy;
+  }
+
+  // Issue read command.
+  uint32_t cmd =
+      mmio_region_read32(otp->base_addr, OTP_CTRL_DIRECT_ACCESS_CMD_REG_OFFSET);
+  cmd = bitfield_bit32_write(cmd, OTP_CTRL_DIRECT_ACCESS_CMD_RD_BIT, 1u);
+  mmio_region_write32(otp->base_addr, OTP_CTRL_DIRECT_ACCESS_CMD_REG_OFFSET,
+                      cmd);
+
+  // Wait for read to complete.
+  do {
+    status = mmio_region_read32(otp->base_addr, OTP_CTRL_STATUS_REG_OFFSET);
+  } while (!bitfield_bit32_read(status, OTP_CTRL_STATUS_DAI_IDLE_BIT));
```

Comment:
Do we need to check any of the error bits here?

Whether or not we'd do detailed error checking for DAI reads depends on whether or not we need DAI reads in the first place.  I might suggest aborting the read with something `kErrorOtpUnknown` if any of the error bits are set in the status register and then leaving a TODO to determine if more comprehensive error reporting is needed.

Reply:
This sounds good. For now I'm checking the `DAI_ERROR` bit and we can add more exhaustive error checking if we plan on using DAI.